File: apps/sdkwork-skills-pc/packages/sdkwork-skills-pc-admin-core/src/permissions.ts
import { isBlank, trim } from '@sdkwork/utils';

export const SKILLS_ADMIN_PERMISSIONS = {
  packageManage: 'skills.packages.manage',
  categoryManage: 'skills.categories.manage',
  marketplaceRead: 'skills.marketplace.read',
  packagesInstall: 'skills.packages.install',
} as const;

export const SKILLS_ADMIN_ROLES = {
  operator: 'skills_admin_operator',
  superAdmin: 'org_admin',
} as const;

export type SkillsAdminPermission =
  (typeof SKILLS_ADMIN_PERMISSIONS)[keyof typeof SKILLS_ADMIN_PERMISSIONS];

export function packageManagePermissionForCategory(categoryCode: string): string {
  return `skills.packages.manage.${categoryCode}`;
}

export function resolveCategoryPackagePermission(category: {
  code: string;
  permission_code?: string | null;
}): string {
  const explicit = category.permission_code ? trim(category.permission_code) : '';
  return !isBlank(explicit) ? explicit : packageManagePermissionForCategory(category.code);
}

export function canManagePackagesInCategories(
  granted: readonly string[],
  roleCodes: readonly string[],
  categoryCodes: readonly string[],
  categories: ReadonlyArray<{ code: string; permission_code?: string | null }> = [],
): boolean {
  if (roleCodes.includes(SKILLS_ADMIN_ROLES.superAdmin)) {
    return true;
  }
  if (granted.includes(SKILLS_ADMIN_PERMISSIONS.packageManage)) {
    return true;
  }
  if (categoryCodes.length === 0) {
    return granted.includes(SKILLS_ADMIN_PERMISSIONS.packageManage);
  }
  const categoryByCode = new Map(categories.map((item) => [item.code, item]));
  return categoryCodes.every((code) =>
    granted.includes(
      resolveCategoryPackagePermission(categoryByCode.get(code) ?? { code }),
    ),
  );
}

File: crates/sdkwork-skills-contract/src/permissions.rs
pub const PERM_MARKETPLACE_READ: &str = "skills.marketplace.read";
pub const PERM_PACKAGES_INSTALL: &str = "skills.packages.install";
pub const PERM_PACKAGES_MANAGE: &str = "skills.packages.manage";
pub const PERM_CATEGORIES_MANAGE: &str = "skills.categories.manage";

pub fn package_manage_permission_for_category(category_code: &str) -> String {
    format!("skills.packages.manage.{category_code}")
}

pub fn resolve_category_package_permission(
    category_code: &str,
    permission_code: Option<&str>,
) -> String {
    permission_code
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
        .unwrap_or_else(|| package_manage_permission_for_category(category_code))
}
